mmdet3d/models/segmentors/encoder_decoder.py

```diff
@@ -0,0 +1,340 @@
+        Then sample points in each patch to batch points of a certain number.
+
+        Args:
+            points (torch.Tensor): Input points of shape [N, 3+C].
+            num_points (int): Number of points to be sampled in each patch.
+            block_size (float, optional): Size of a patch to sample.
+            sample_rate (float, optional): Stride used in sliding patch.
+                Defaults to 0.5.
+            use_normalized_coord (bool, optional): Whether to use normalized \
+                xyz as additional features. Defaults to False.
+
+        Returns:
+            np.ndarray | np.ndarray:
+
+                - patch_points (torch.Tensor): Points of different patches of \
+                    shape [K, N, 3+C].
+                - patch_idxs (torch.Tensor): Index of each point in \
+                    `patch_points`, of shape [K, N].
+        """
+        device = points.device
+        # we assume the first three dims are points' 3D coordinates
+        # and the rest dims are their per-point features
+        coords = points[:, :3]
+        feats = points[:, 3:]
+
+        coord_max = coords.max(0)[0]
+        coord_min = coords.min(0)[0]
+        stride = block_size * sample_rate
+        num_grid_x = int(
+            torch.ceil((coord_max[0] - coord_min[0] - block_size) /
+                       stride).item() + 1)
+        num_grid_y = int(
+            torch.ceil((coord_max[1] - coord_min[1] - block_size) /
+                       stride).item() + 1)
+
+        patch_points, patch_idxs = [], []
+        for idx_y in range(num_grid_y):
+            s_y = coord_min[1] + idx_y * stride
+            e_y = torch.min(s_y + block_size, coord_max[1])
+            s_y = e_y - block_size
+            for idx_x in range(num_grid_x):
+                s_x = coord_min[0] + idx_x * stride
+                e_x = torch.min(s_x + block_size, coord_max[0])
+                s_x = e_x - block_size
+
+                # extract points within this patch
+                cur_min = torch.tensor([s_x, s_y, coord_min[2]]).to(device)
+                cur_max = torch.tensor([e_x, e_y, coord_max[2]]).to(device)
+                cur_choice = ((coords >= cur_min - 1e-3) &
+                              (coords <= cur_max + 1e-3)).all(dim=1)
+
+                if not cur_choice.any():  # no points in this patch
+                    continue
+
+                # sample points in this patch to multiple batches
+                cur_center = cur_min + block_size / 2.0
+                point_idxs = torch.nonzero(cur_choice, as_tuple=True)[0]
+                num_batch = int(np.ceil(point_idxs.shape[0] / num_points))
+                point_size = int(num_batch * num_points)
+                replace = point_size > 2 * point_idxs.shape[0]
+
+                # TODO: torch alternative for np.random.choice?
+                # TODO: tried tensor.multinomial before, finding it very slow
+                point_idxs_repeat = torch.from_numpy(
```

Comment:
torch.randperm()[:num]

Reply:
The main difficulty here is that, we don't know whether `replace=True/False`. If `replace=False`, I don't have a simple way to perform random.choice using `torch`?

Reply:
Use pure torch code.